x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/RollupILMAction.java

```diff
@@ -0,0 +1,131 @@
+ */
+public class RollupILMAction implements LifecycleAction {
+    public static final String NAME = "rollup";
+
+    private static final ParseField CONFIG_FIELD = new ParseField("config");
+    private static final ParseField POLICY_FIELD = new ParseField("rollup_policy");
+
+    @SuppressWarnings("unchecked")
+    private static final ConstructingObjectParser<RollupILMAction, Void> PARSER = new ConstructingObjectParser<>(NAME,
+        a -> new RollupILMAction((RollupActionConfig) a[0], (String) a[1]));
+
+    private final RollupActionConfig config;
+    private final String rollupPolicy;
+
+    static {
+        PARSER.declareField(ConstructingObjectParser.constructorArg(),
+            (p, c) -> RollupActionConfig.fromXContent(p), CONFIG_FIELD, ObjectParser.ValueType.OBJECT);
+        PARSER.declareString(ConstructingObjectParser.optionalConstructorArg(), POLICY_FIELD);
+    }
+
+    public static RollupILMAction parse(XContentParser parser) {
+        return PARSER.apply(parser, null);
+    }
+
+    public RollupILMAction(RollupActionConfig config, @Nullable String rollupPolicy) {
+        this.config = config;
+        this.rollupPolicy = rollupPolicy;
+    }
+
+    public RollupILMAction(StreamInput in) throws IOException {
+        this(new RollupActionConfig(in), in.readOptionalString());
+    }
+
+    @Override
+    public String getWriteableName() {
+        return NAME;
+    }
+
+    RollupActionConfig config() {
+        return config;
+    }
+
+    String rollupPolicy() {
+        return rollupPolicy;
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject();
+        builder.field(CONFIG_FIELD.getPreferredName(), config);
+        if (rollupPolicy != null) {
+            builder.field(POLICY_FIELD.getPreferredName(), rollupPolicy);
+        }
+        builder.endObject();
+        return builder;
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        config.writeTo(out);
+        out.writeOptionalString(rollupPolicy);
+    }
+
+    @Override
+    public boolean isSafeAction() {
+        return false;
+    }
+
+    @Override
+    public List<Step> toSteps(Client client, String phase, StepKey nextStepKey) {
+        StepKey checkNotWriteIndex = new StepKey(phase, NAME, CheckNotDataStreamWriteIndexStep.NAME);
+        StepKey readOnlyKey = new StepKey(phase, NAME, ReadOnlyAction.NAME);
+        StepKey rollupKey = new StepKey(phase, NAME, NAME);
+        Settings readOnlySettings = Settings.builder().put(IndexMetadata.SETTING_BLOCKS_WRITE, true).build();
+        CheckNotDataStreamWriteIndexStep checkNotWriteIndexStep = new CheckNotDataStreamWriteIndexStep(checkNotWriteIndex,
+            readOnlyKey);
+        UpdateSettingsStep readOnlyStep = new UpdateSettingsStep(readOnlyKey, rollupKey, client, readOnlySettings);
```

Comment:
This should use the `ReadOnlyStep` (which uses the new API for marking things as read only) rather than updating the settings manually